### PR TITLE
fix(csharp): qualified/alias constructor names, : base/: this initializers, generic type-arg strip

### LINKED ISSUE #1924: C#: parsing-layer coverage gaps (7 findings)
**Parent epic:** #1919
**Enabled by:** #1920 (node-type validator) — closes F15 · #1921 (heritage helper) — closes F10, F11

7 verified findings (1 HIGH). Each box is a discrete fix unit. Legend: ✅ confirmed · 🟡 partial.

- [ ] ✅ **F10 [HIGH]** record and struct heritage (inheritance / interface implementation) never captured
  - _grammar:_ node-types.json: record_declaration and struct_declaration each have 'base_list' in their children subtypes (both single, named=true) — confirmed present.
  - _source:_ tree-sitter-queries.ts:885-897 heritage block only matches (class_declaration ... base_list) and (interface_declaration ... base_list)
- [ ] ✅ **F11 [MEDIUM]** Qualified-name and primary-constructor base types are not captured for heritage
- [ ] ✅ **F12 [MEDIUM]** Alias `using` directive emits the alias name as @import.source (spurious import) in the legacy query path
- [ ] ✅ **F13 [MEDIUM]** this./base. field writes are not captured by the central assignment write pattern
- [ ] ✅ **F14 [LOW]** Alias using to a non-qualified or generic target produces no named binding
- [ ] ✅ **F15 [LOW]** Dead grammar-shape branches: equals_value_clause and identifier_name node types do not exist in 0.23.1
- [ ] ✅ **F16 [LOW]** receiver-binding `base` type text includes constructor args for C#12 primary-constructor bases

_Verified against `main` @ `a93ecee0`. Finding IDs use an `F` prefix to avoid colliding with unrelated issue/PR numbers._

### PR DESCRIPTION
**Pull request by Sweetdevil144**
## Summary

The three Java parsing-layer defects fixed in #1928 exist **verbatim** in the C# scope-resolution path. This applies the same fixes for C#:

- **Qualified / qualified-generic / alias-qualified constructor calls.** `new Ns.Foo()`, `new A.B.Foo()`, `new Ns.Box<int>()` bound only `@reference.call.constructor.qualified` with **no** `@reference.name`, so the central extractor fell back to the whole-expression anchor and the reference name became the raw `new Ns.Foo()` text (never resolved). The simple-name tail is now derived via the existing `terminalTypeNameNode` helper (handles `qualified_name`, a generic tail, and `alias_qualified_name`). **Also closes a true coverage gap:** top-level `alias_qualified_name` constructors (`new MyAlias::Foo()`, `new global::Foo()`) matched **no** query arm and produced no reference at all — a new arm captures them.
- **`: base(...)` / `: this(...)` explicit constructor initializers.** Modeled by tree-sitter as `constructor_initializer`, never matched by the scope query, so chained-constructor `CALLS` edges were dropped. Synthesized with the target resolved structurally — `this` → enclosing type name; `base` → the base type's bare name (the first base-list entry, which C# requires to be the base class). Arity attached for overload disambiguation. Covers class/struct/record.
- **`interpretCsharpTypeBinding` qualifier strip.** Used `lastIndexOf('.')` over the whole string, cutting inside a qualified generic **type argument** (`Dictionary<string, Ns.User>` → `User>`). `stripQualifier` is now generic-aware: it reduces only the segment before the first `<` and re-attaches the generic suffix, so multi-arg generics stay intact and the `.Values`/`.Keys` collection-accessor unwrap keeps working.

## Test plan

- [x] `test/unit/scope-resolution/csharp/csharp-captures.test.ts` — every constructor shape (identifier, generic, qualified, nested, qualified-generic, **alias-qualified**, alias-then-qualified), double-match guard; `: base`/`: this` incl. struct/record/**mixed base list** (base class first per C# rules) and the no-base-list defensive case.
- [x] `test/unit/scope-resolution/csharp/csharp-interpret.test.ts` — the `Dictionary<string, Ns.User>` corruption case, unqualified multi-arg (regression guard), outer-qualified generic, single-arg element extraction, nullable, **nullable+generic**, **nested generic**, **unknown qualified generic**, collection-accessor suffix.
- [x] `test/integration/resolvers/csharp-1924.test.ts` (+ `csharp-qualified-constructor` / `csharp-explicit-ctor-init` fixtures) — asserts the `CALLS` edges resolve and no corrupted qualified name leaks.
- [x] Full C# suite (331) green, `tsc --noEmit` clean. The `csharp-captures` golden was regenerated; drift is **purely additive** (only the new fixtures — zero existing-fixture digests changed).

> Note: my first `stripQualifier` attempt (blanket generic erasure) broke 3 `Dictionary<…>.Values` unwrap tests; the existing suite caught it and the fix was narrowed to the generic-aware approach above.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: gitnexus/src/core/ingestion/languages/csharp/captures.ts [hunk 1]
```diff
@@ -263,10 +282,93 @@ export function emitCsharpScopeCaptures(
...
+ * query never matched, so the chained-constructor CALLS edges (derived ctor →
+ * base ctor; ctor → sibling overload) were silently dropped.
+ *
+ * The initializer carries no constructor name (the `base`/`this` child is a bare
+ * keyword token), so the target is resolved structurally:
+ *   - `this(...)` → the enclosing type's own simple name.
+ *   - `base(...)` → the enclosing class/record's base type, reduced to its bare
+ *     simple name via `terminalTypeNameNode`. C# base lists do not syntactically
+ *     separate the base class from interfaces; the base class is idiomatically
+ *     first, and an interface target simply finds no constructor (no edge) rather
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/csharp/captures.ts]: **[P2] This stated guarantee is false — `: base()` on an interface-only class emits a real CALLS edge to the interface.** `class C : IFoo, IBar { C() : base() {} }` is valid C# (base is implicit `System.Object`). The base-list loop just below picks the first non-builtin entry (`IFoo`); downstream, `walkers.ts` `isClassLike` includes `Interface` and `pickConstructorOrClass` (free-call-fallback.ts) returns the interface node when it finds no constructor — so the edge is emitted, not suppressed.

_Reproduced_ (`REGISTRY_PRIMARY_CSHARP=1`): observed `Constructor:C.C#0 → Interface:IFoo`.

Fix: skip the `base` reference when the chosen entry resolves to an `Interface`-kind def (or have the resolver suppress constructor edges whose target is an interface), and correct this comment. `[reproduced]`

File: gitnexus/src/core/ingestion/languages/csharp/captures.ts [hunk 2]
```diff
@@ -263,10 +282,93 @@ export function emitCsharpScopeCaptures(
...
+    const argList = findNamedChild(node, 'argument_list');
+    const arity =
+      argList === null
+        ? 0
+        : argList.namedChildren.filter((c) => c !== null && c.type === 'argument').length;
+
+    out.push({
+      '@reference.call.constructor': nodeToCapture('@reference.call.constructor', node),
+      '@reference.name': nodeToCapture('@reference.name', targetNameNode),
+      '@reference.arity': syntheticCapture('@reference.arity', node, String(arity)),
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/csharp/captures.ts:365]: **[P2] Synthesized arity is cosmetic for constructor targets → multi-overload `: this(...)`/`: base(...)` produce wrong-overload & self-loop edges.** `pickConstructorOrClass` (free-call-fallback.ts) returns the *first* Constructor and never reads `site.arity`; and because every ctor def is auto-hoisted to the class scope (scope-extractor.ts ~528), `resolveCallerGraphId` (graph-bridge/ids.ts:201) also sources the edge from the *first* ctor.

_Reproduced_: `class P { P(int a):this(a,0){} P(int a,int b){} }` → self-loop `P#1→P#1` (correct `P#1→P#2` missing); `class Q { Q(){} Q(int a):this(){} }` → edge sourced from `Q#0` though `:this()` is in `Q#1`.

Fix: narrow constructor targets by `site.arity` (reuse `narrowByArity`) and resolve the initializer's caller to the specific enclosing constructor, not the class's first. `[reproduced]`

File: gitnexus/src/core/ingestion/languages/csharp/captures.ts [hunk 3]
```diff
@@ -196,6 +196,25 @@ export function emitCsharpScopeCaptures(
...
+    // `new Ns.Foo()` text (never resolves). Derive the bare simple-name tail via
+    // the same `terminalTypeNameNode` helper the inheritance synth uses — it
+    // handles qualified_name, a generic tail (`Ns.Box<int>` → `Box`), and
+    // alias_qualified_name (`global::Ns.Foo`). Mirrors Java F35 (#1928).
+    if (
+      grouped['@reference.call.constructor.qualified'] !== undefined &&
+      grouped['@reference.name'] === undefined
+    ) {
+      const qNode = nodeMap['@reference.call.constructor.qualified'];
+      const nameNode = qNode === undefined ? null : terminalTypeNameNode(qNode);
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/csharp/captures.ts:212]: **[P2] Dropping the qualifier here makes a *qualified* `new B.Foo()` mis-bind across namespaces.** Reducing to the bare tail with no `@reference.qualified-name` means resolution is simple-name + local-scope-first.

_Reproduced_: `App` in `namespace A` doing `new B.Foo()`, with both `A.Foo` and `B.Foo` present, binds to **`A.Foo`** (target id `Foo.Foo#0` in A's file). The alias form `new Lib::Foo()` likewise discards the alias. Before this PR a qualified `new B.Foo()` produced *no* edge; now it can produce a *wrong* one — the #1991/#1993 same-tail hazard, now live for ctors.

Fix: thread the discarded qualifier as `@reference.qualified-name` and add a qualified-first lookup for constructor calls (mirroring `resolveInheritanceBaseInScope`). `[reproduced]`

File: gitnexus/src/core/ingestion/languages/csharp/interpret.ts [hunk 4]
```diff
@@ -129,9 +129,17 @@ function stripGeneric(text: string): string {
...
+  // the segment before the first `<`; re-attach the generic suffix unchanged.
+  const ltIdx = text.indexOf('<');
+  const base = ltIdx === -1 ? text : text.slice(0, ltIdx);
+  const lastDot = base.lastIndexOf('.');
   if (lastDot === -1) return text;
-  const tail = text.slice(lastDot + 1);
+  const tail = base.slice(lastDot + 1);
   if (COLLECTION_ACCESSOR_SUFFIXES.has(tail)) return text;
-  return tail;
+  return ltIdx === -1 ? tail : tail + text.slice(ltIdx);
```
Comment (magyargergo) [inline gitnexus/src/core/ingestion/languages/csharp/interpret.ts]: **[P3] This re-attach narrows the F41 fix but regresses nested types through a generic outer.** Slicing at the *first* `<` and re-appending everything after it means a nested type whose outer is generic keeps the whole suffix.

_Reproduced_: `Ns.Outer<int>.Inner` → `"Outer<int>.Inner"` (unresolvable), where the old `lastIndexOf('.')` gave the bare `Inner`. (Non-generic `Outer.Inner` → `Inner` still works.) Silent dropped binding for that shape.

Fix: find the last `.` at angle-bracket depth 0 (track `<`/`>` depth) rather than slicing at the first `<` — keeps the F41 type-argument fix *and* preserves the `Outer<int>.Inner → Inner` reduction. `[reproduced]`

File: gitnexus/test/integration/resolvers/csharp-1924.test.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    const calls = getRelationships(result, 'CALLS');
+    const baseCall = calls.find((c) => c.target === 'Base');
+    expect(baseCall).toBeDefined();
+    expect(baseCall!.source).toBe('Child');
+    expect(['Class', 'Constructor']).toContain(baseCall!.targetLabel);
+  });
+
+  it('resolves `: this()` to the sibling Child constructor', () => {
+    const calls = getRelationships(result, 'CALLS');
+    const thisCall = calls.find((c) => c.target === 'Child' && c.source === 'Child');
```
Comment (magyargergo) [inline gitnexus/test/integration/resolvers/csharp-1924.test.ts:61]: **[P2] This assertion masks the self-loop bug above.** `c.target === 'Child' && c.source === 'Child'` is satisfied by *both* the correct `Child(int)→Child()` edge and the buggy `Child()→Child()` self-loop — it matches on name only and never checks `sourceLabel`/node identity.

Strengthen it to pin the edge endpoints by node id/label (e.g. assert the source is the `Child(int)` constructor and the target is the `Child()` overload), and add a negative test for an interface-only base list (`class C : IFoo` + `: base()` asserting **no** CALLS edge to `IFoo`). Also missing: an end-to-end alias-ctor (`new Alias::Type()`) test. `[reproduced]`

**Review by magyargergo — COMMENTED** [review]
# PR #2046 Tri-Review — C# qualified/initializer constructor resolution

**Methods (engine breakdown):** 3 methods. **Codex** (the one independent engine) ran LIVE (foreground retry after the async wrapper evaporated). **8 Claude lanes** under different personas: gitnexus risk + test/CI, CE correctness/adversarial/performance/testing, plus the three you requested — static-analysis (tree-sitter/AST), compiler/interpreter (type-normalization), and knowledge-graph builder (edge correctness). Two of three methods are Claude, so Claude-only agreement is "consistent across personas," not independent confirmation. The headline findings below were **reproduced end-to-end by the coordinator** against the live pipeline (`REGISTRY_PRIMARY_CSHARP=1`), the strongest tier.

## What's solid (validated)
- **F35 (qualified/alias/global/generic ctor names) is correct.** `terminalTypeNameNode` reductions for `qualified_name`, `generic_name`, `alias_qualified_name`, `primary_constructor_base_type` were verified against a real tree-sitter-c-sharp parse. `new global::Ns.Foo()` (nests under `qualified_name`) vs `new MyAlias::Foo()` (alias branch) routing is correct; no double-capture.
- **F41 `stripQualifier` is a real bug fix.** It previously corrupted `Dictionary<string, Ns.User>` → `User>`, breaking the `.Values`/`.Keys` accessor unwrap. The rewrite fixes it; verified across the unit corpus.
- **No O(n²) regression.** The added `synthesizeCsharpConstructorInitializerReferences` is one extra linear `walkNamedTree` pass (2→3 full-tree walks); `findEnclosingTypeDeclaration` is bounded by nesting depth; the F35 block is O(1) per match via `nodeMap` (no root-walk). Clean.
- **Worker-mode parity is intact.** Worker mode calls `extractParsedFile(provider, …)` → `provider.emitScopeCaptures` = `emitCsharpScopeCaptures` (single source of truth; csharp.ts:202). The fix runs identically in both paths — no parse-worker copy to keep in lockstep.

## Headline findings (all coordinator-reproduced)
The **F38 constructor-initializer** feature is correct for the common case (single ctor, clean `class : Base`) but has graph-correctness gaps in multi-overload / interface-only edge cases — and the PR's own tests mask them.

1. **[P2] Interface-only `: base()` emits a FALSE CALLS edge to an interface.** `class C : IFoo, IBar { C() : base() {} }` is valid C# (base is implicit `System.Object`). The synth picks the first non-builtin base-list entry = `IFoo`; the resolver's `isClassLike` includes `Interface` and `pickConstructorOrClass` returns the interface node when it finds no constructor → `Constructor:C → Interface:IFoo`. The code comment at `captures.ts:303` claims this is impossible ("simply finds no constructor (no edge)") — that claim is **factually wrong**. Reproduced: edge `C.C#0 → Interface:IFoo`. *(Codex + adversarial + static-analysis + knowledge-graph + reproduced.)*

2. **[P2] Multi-overload `: this(...)` / `: base(...)` produce wrong-overload and self-loop edges.** The synthesized `@reference.arity` is **cosmetic for constructor targets**: `pickConstructorOrClass` (free-call-fallback.ts) returns the *first* Constructor and never consults arity, and `resolveCallerGraphId` (graph-bridge/ids.ts:201, `ownedDefs.find(d => Constructor)`) attributes the caller to the *first* Constructor of a multi-ctor class. The reason both ctors share one `ownedDefs`: although `(constructor_declaration) @scope.function` gives each ctor its own scope, `scope-extractor.ts` auto-hoists scope-creating declarations' defs to the **parent (class) scope** (`rangesEqual(anchor.range, innermost.range)`, ~line 528), so the class scope's `ownedDefs` holds *all* ctors and `find` returns the first. (This is why a purely-theoretical reading — "each ctor has its own scope, so the caller must be right" — is wrong; the reproduction settles it.) Reproduced: `class P { P(int a):this(a,0){} P(int a,int b){} }` → self-loop `P#1→P#1` (correct `P#1→P#2` missing); `class Q { Q(){} Q(int a):this(){} }` → edge sourced from `Q#0` though `:this()` is written in `Q#1`. *(Codex + knowledge-graph + reproduced. The underlying arity-blindness/first-ctor selection is a pre-existing resolver property, but this PR newly turns initializers into call sites that trigger it.)*

3. **[P2] Qualifier is dropped, so a *qualified* `new B.Foo()` mis-binds across namespaces.** F35 reduces to the bare simple name and emits no `@reference.qualified-name`, so resolution is simple-name + local-scope-first. Reproduced: `App` in `namespace A` doing `new B.Foo()`, with both `A.Foo` and `B.Foo` present, binds to **`A.Foo`** (wrong). The alias form `new Lib::Foo()` likewise discards the alias (Codex NF1, same root cause). Before this PR a qualified `new B.Foo()` produced *no* edge; now it can produce a *wrong* one. This is the #1991/#1993 same-tail hazard, live for constructor calls. *(Codex + knowledge-graph + reproduced. Unqualified `new Foo()` resolving by simple name is the established contract; the new aspect is ignoring an explicit `B.`/alias qualifier.)*

4. **[P3] `stripQualifier` narrows F41 but introduces a small regression for nested types through a generic outer.** `Ns.Outer<int>.Inner` now returns `Outer<int>.Inner` (unresolvable) where the old code returned the bare `Inner`. Reproduced: `raw('Ns.Outer<int>.Inner') === 'Outer<int>.Inner'` (non-generic `Outer.Inner` → `Inner` still works). Narrow/uncommon shape, silent dropped-edge. Fix: find the last `.` at angle-bracket depth 0 instead of slicing at the first `<` — keeps the F41 fix and preserves the nested-type reduction. *(Codex + correctness + adversarial + reproduced.)*

5. **[P2] The new tests mask findings 1–2.** `csharp-1924.test.ts:61` asserts `:this()` produces an edge with `source==='Child' && target==='Child'` — which the self-loop bug satisfies. No assertion checks `sourceLabel` or node identity, so a correct `Child(int)→Child()` edge and the buggy `Child()→Child()` self-loop are indistinguishable. There is also no interface-only-base negative test and no end-to-end alias-ctor test.

## Lower-priority / notes
- **Test coverage narrowed (mixed).** `csharp-captures.test.ts` was rewritten (-479 lines). The deleted **adaptive-buffer / UTF-8 large-file** tests are genuinely redundant (covered by `ingestion-utils.test.ts` + `csharp-hooks.test.ts:302` + integration `csharp.test.ts:2501` / issue #1066) — fine. But the deleted **operator / conversion-operator / destructor / `params`-variadic / receiver-binding** assertions have **no backstop**: no `lang-resolution/csharp-*` fixture contains `operator`, `~`, or `params`, so the golden-corpus digest can't catch drift there. Those capture behaviors are now unguarded. *(testing + correctness + static-analysis.)*
- **Refuted suspicions (validation):** no duplicate constructor edge (the F35 path mutates the existing match in place; `handledSites` dedups vs the lookup pass); target-typed `new()` / `new int[]` / `new {}` produce zero constructor matches (no crash, no edge); `terminalTypeNameNode` is null-safe on every probed shape (`object`/predefined types skipped); the F41 fix correctly preserves multi-arg generics.

## CI
- **Vercel: fail** — "Authorization required to deploy" (infra/preview-deploy auth, not a code failure). Apply-conventional-label: pass. Title check: skipped. Branch is `BEHIND` main (mergeable).
- **[P2] The tests that would catch findings 1–3 likely do not run in CI.** They are resolver integration tests that **skip unless `REGISTRY_PRIMARY_CSHARP=1`** is set in the environment (repo convention). No check in this PR shows that suite running. Combined with finding #5 (the assertions that do run are too loose to distinguish a self-loop from a correct edge), the constructor-resolution behavior is effectively unguarded in CI. Worth confirming the resolver matrix actually exercises `csharp-1924.test.ts` with the env flag set — otherwise these edges regress silently.

---
*Automated multi-tool digest (Codex live + 8 Claude lanes; headlines coordinator-reproduced). Findings 1–5 reproduced end-to-end; verify before acting. Severity is the coordinator's calibration — findings 2 and 4 are narrow edge cases (multi-overload ctors / nested-generic annotations), while finding 3 (cross-namespace same-name `new B.Foo()`) is more plausible in large multi-namespace C# codebases. None is common-path breakage of the single-ctor clean-base case.*
